### PR TITLE
#2435 - Make modifier_order rule rely on an explicit set of rules 

### PR DESCRIPTION
**Pull request by biboran**
After a bit of research resulting from #2435 it was found that `modifier_order` rule only cares about specified exact leading ordering. For example given the ordering `[.final, .override, .acl]` the following would be considered triggering:

```swift
weak final override private var a: String
final weak override private var b: String
final override weak private var c: String
```
but this would be considered fine:
```swift
final override private weak var d: String
```

Such behavior is not always desired. It might be more beneficial if `modifier_order` rule only enforced order between the specified modifiers and left the unspecified in place. I slightly modified the algorithm and adjusted it such that it's a bit stricter than the current `oss-check`.

### CODE REVIEW

File: Source/SwiftLintFramework/Rules/Style/ModifierOrderRule.swift [hunk 1]
```diff
@@ -1,10 +1,21 @@
 import SourceKittenFramework
 
-private typealias ModifierKeyword = String
-private typealias ModifierDescription = (ModifierKeyword, SwiftDeclarationAttributeKind.ModifierGroup)
-
 public struct ModifierOrderRule: ASTRule, OptInRule, ConfigurationProviderRule {
-    public var configuration = ModifierOrderConfiguration(preferredModifierOrder: [.override, .acl])
+    public var configuration = ModifierOrderConfiguration(
+        preferredModifierOrder: [
```
[the remark this answers is not in the document]
Comment (biboran) [reply Source/SwiftLintFramework/Rules/Style/ModifierOrderRule.swift]: Leaving the order as is would fix all of the violations where some other modifier preceded or was in-between the `[.override, .acl]` modifiers. Matching the prior behavior of the rule with this version was impossible so I just used the `oss-check` as a guideline to establish the modifier order. If this change is desired the order should definitely be made less strict.

File: Source/SwiftLintFramework/Rules/Style/ModifierOrderRule.swift [hunk 2]
```diff
@@ -14,163 +25,8 @@ public struct ModifierOrderRule: ASTRule, OptInRule, ConfigurationProviderRule {
         description: "Modifier order should be consistent.",
         kind: .style,
         minSwiftVersion: .fourDotOne ,
-        nonTriggeringExamples: [
```
[the remark this answers is not in the document]
Comment (biboran) [reply Source/SwiftLintFramework/Rules/Style/ModifierOrderRule.swift]: Sure, moved to a separate commit. None of the test cases were changed

[the remark this answers is not in the document]
Comment (masters3d) [reply Source/SwiftLintFramework/Rules/Style/ModifierOrderRule.swift]: Can you add some non triggering test cases like the ones you described above?

[the remark this answers is not in the document]
Comment (biboran) [reply Source/SwiftLintFramework/Rules/Style/ModifierOrderRule.swift]: Added the examples

**Comment by jpsim** [comment]
Thank you very much for this PR @biboran! I had concerns about this when the rule was originally added, so it's nice to see you address them 😄 

I made some small changes and added a changelog entry. I'll merge when CI is happy.

**Pull request merged**